collector/collector.go

```diff
@@ -512,3 +521,27 @@ func modifyNumericMetric(data interface{}, scale float64, shift float64) interfa
 	}
 	return modifiedData
 }
+
+func snmpResults2Map(results []*snmpgo.VarBind, oid string) (map[int]*snmpgo.VarBind, serror.SnapError) {
+	index := ""
+	res := make(map[int]*snmpgo.VarBind)
+	oidTrimmed := strings.Trim(oid, ".")
+	oidLen := len(strings.Split(oidTrimmed, "."))
+
+	for _, r := range results {
+		roidTrimmed := strings.Trim(r.Oid.String(), ".")
+		roidLen := len(strings.Split(roidTrimmed, "."))
+		if ((roidLen == oidLen) || (roidLen == oidLen+1)) && strings.HasPrefix(roidTrimmed, oidTrimmed) {
```

Comment:
@michep Thanks for detailed explanation of this condition.

Please notice that for both queries you have the same OID in response
```
# snmpwalk -v 2c -On -c public 10.0.0.4 .1.3.6.1.2.1.1.5
.1.3.6.1.2.1.1.5.0 = STRING: mfms_switch1

# snmpwalk -v 2c -On -c public 10.0.0.4 .1.3.6.1.2.1.1.5.0
.1.3.6.1.2.1.1.5.0 = STRING: mfms_switch1```

r.Oid.String() is the OID from device response so it should have the same form and it doesn't depend on OID in SNMP request.

Why in this function `oid string` is needed as an argument of `snmpResults2Map`? Taking into account that device replies with OID in full form (for request with .1.3.6.1.2.1.1.5 and .1.3.6.1.2.1.1.5.0 in response there is .1.3.6.1.2.1.1.5.0 as OID), you can split `roidTrimmed` and you use the last element as `index`, so the if condition is not needed.

It is what I see but please correct me if I'm wrong and you catch response from device in different form.